File: q26-50/q31.py
# 2020.08.03
# Problem Statement:
# https://leetcode.com/problems/next-permutation/

class Solution:
    # check if the list is already the largest
    def isLargest(nums):
        for i in range(0, len(nums)-1):
            if nums[i] < nums[i+1]:
                return False
        return True
    
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # check empty list
        if len(nums) == 0: return
        
        # check if the list is already the largest
        if Solution.isLargest(nums):
            nums.sort()
            return
        
        # idea:
        # search from right to the left
        # swap the first number which violets the ascending order (from right to left)
        # with the (smallest, but also larger than it) number to its right
        # sort the right part in ascending order

        # initialize the first index to change
        index_to_change_1 = 0
        
        # find the first violation index
        for i in range(0, len(nums)-1):
            if nums[len(nums)-1-i] > nums[len(nums)-1-i-1]:
                index_to_change_1 = len(nums)-1-i-1
                break                
        num_to_change_1 = nums[index_to_change_1]

        # initialize the info of the other number to swap
        num_to_change_2, index_to_change_2 = 0, 0

        # seach for the number (the smallest number with is larger than num_to_change_1)
        diff = inf
        for number in nums[index_to_change_1+1: len(nums)]:
            if number > num_to_change_1 and number-num_to_change_1 < diff:
                diff = number-num_to_change_1
                num_to_change_2 = number
                index_to_change_2 = index_to_change_1+1 \
                + nums[index_to_change_1+1: len(nums)].index(number)
        
        # do the swap
        nums[index_to_change_1] = num_to_change_2
        nums[index_to_change_2] = num_to_change_1
        
        # sort the rest
        nums[index_to_change_1+1: len(nums)] = sorted(nums[index_to_change_1+1: len(nums)])

        return

File: Amazon/q31.java
// 2022.05.28
// Problem Statement:
// https://leetcode.com/problems/next-permutation/

// idea: if last 2 is decreasing, directly swap,
// find the last index that starts increasing till the end, 
// and find the number in the increasing subarray that's slightly larger
// than the previous (of the increasing subarray) number,
// swap the previous number and the found number, 
// reverse the increasing subarray
class Solution {
    public void nextPermutation(int[] nums) {
        if (nums.length==1) return;
        if (nums[nums.length-1]>nums[nums.length-2]) {
            int temp = nums[nums.length-1];
            nums[nums.length-1] = nums[nums.length-2];
            nums[nums.length-2] = temp;
            return;
        }
        int idx = nums.length-1;
        for (int i=nums.length-1; i>=1; i--) {
            idx = i;
            if (nums[i]>nums[i-1]) { // increasing
                break;
            }
        }
        
        // 0 to idx-1: unknown
        // idx till end: decreasing
        if (idx==1 && nums[0]>=nums[1]) { // decreasing from the beginning
            for (int i=0; i<nums.length/2; i++) {
                int j = nums.length-1-i;
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
            }
            return;
        }
        
        // find the number that's slightly larger than nums[idx-1],
        // set nums[idx-1] = that number,
        // put nums[idx-1] to nums[idx:]
        for (int i=nums.length-1; i>=idx; i--) {
            if (nums[i]>nums[idx-1]) {
                int temp = nums[idx-1];
                nums[idx-1] = nums[i];
                nums[i] = temp;
                break;
            }
        }
        
        // then reverse
        for (int i=idx; i<(nums.length+idx)/2; i++) {
            int j = nums.length-1-(i-idx);
            int temp = nums[i];
            nums[i] = nums[j];
            nums[j] = temp;
        }
        return;
    }
}